File: LeetCode/Medium/1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ret = 0;
    int longestZigZag(TreeNode* root) {
        DFS(root, 0, 0);
        return ret;
    }
    
    void DFS(TreeNode* node, int left, int right){
        if (!node) return;
        
        ret = max(ret, max(left, right));
        
        DFS(node->left, right+1, 0);
        DFS(node->right, 0, left+1);
    }
};

File: LeetCode/Medium/1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ret = 0;
     public int longestZigZag(TreeNode root) {
        DFS(root, true, 0);
        DFS(root, false, 0);
        return this.ret - 1;
    }
    
    private void DFS(TreeNode node, boolean is_left, int ret){
        if (node == null) {
            this.ret = Math.max(ret, this.ret);
            return;
        }
        DFS(node.left, !is_left, is_left ? ret + 1 : 0);
        DFS(node.right, !is_left, is_left ? 0 : ret + 1);
    }
}